Add resource metrics to Kafka describe cluster response

MikeEdgar opened the pull request:
Builds on #166, see only latest commit in this PR for metrics changes.

Queries from Prometheus and adds time series for the following metrics: pod CPU usage, pod memory usage, single Kafka listener incoming byte rate, single Kafka listener outgoing byte rate, persistent volume capacity and usage (when PVC is used by Kafka nodes).

`GET http://localhost:8080/api/kafkas/FZdRtosjRMyJI6aa9noffQ?fields[kafkas]=name,metrics`

<details><summary>Sample Response</summary>
<p>

```json
{
  "data": {
    "id": "FZdRtosjRMyJI6aa9noffQ",
    "type": "kafkas",
    "attributes": {
      "name": "kafka2",
      "metrics": {
        "cpu_usage_seconds": {
          "0": [
            [
              1699653178.33,
              "0.12229363795710607"
            ],
            [
              1699653238.33,
              "0.14317560650523886"
            ],
            [
              1699653298.33,
              "0.12029969629153833"
            ],
            [
              1699653358.33,
              "0.13543602508798774"
            ],
            [
              1699653418.33,
              "0.12852764913844292"
            ],
            [
              1699653478.33,
              "0.11419059980294952"
            ],
            [
              1699653538.33,
              "0.12539510361956716"
            ],
            [
              1699653598.33,
              "0.1238172762757376"
            ],
            [
              1699653658.33,
              "0.1338190300934218"
            ],
            [
              1699653718.33,
              "0.13333526357676329"
            ],
            [
              1699653778.33,
              "0.12536614638839255"
            ],
            [
              1699653838.33,
              "0.16099633520954315"
            ],
            [
              1699653898.33,
              "0.11871236782341218"
            ],
            [
              1699653958.33,
              "0.13966191337081096"
            ],
            [
              1699654018.33,
              "0.11309469830573524"
            ],
            [
              1699654078.33,
              "0.1432740337564133"
            ],
            [
              1699654138.33,
              "0.1555230959931351"
            ],
            [
              1699654198.33,
              "0.1185391204286054"
            ],
            [
              1699654258.33,
              "0.1552563051458573"
            ],
            [
              1699654318.33,
              "0.1216215466211638"
            ],
            [
              1699654378.33,
              "0.14048110866529434"
            ],
            [
              1699654438.33,
              "0.11576054621054407"
            ],
            [
              1699654498.33,
              "0.14119084836546897"
            ],
            [
              1699654558.33,
              "0.11411384590749644"
            ],
            [
              1699654618.33,
              "0.10945637039838017"
            ],
            [
              1699654678.33,
              "0.09906739616159926"
            ],
            [
              1699654738.33,
              "0.14022924306479603"
            ],
            [
              1699654798.33,
              "0.14675280201284988"
            ],
            [
              1699654858.33,
              "0.11415181745289095"
            ],
            [
              1699654918.33,
              "0.14629864406748028"
            ],
            [
              1699654978.33,
              "0.17798586759932727"
            ]
          ],
          "1": [
            [
              1699653178.33,
              "0.12460929422349869"
            ],
            [
              1699653238.33,
              "0.11928209482940874"
            ],
            [
              1699653298.33,
              "0.13299701940161346"
            ],
            [
              1699653358.33,
              "0.12396957026550304"
            ],
            [
              1699653418.33,
              "0.12985607489226247"
            ],
            [
              1699653478.33,
              "0.11364991388721704"
            ],
            [
              1699653538.33,
              "0.12950766497600005"
            ],
            [
              1699653598.33,
              "0.1347759207945597"
            ],
            [
              1699653658.33,
              "0.14284475258117166"
            ],
            [
              1699653718.33,
              "0.1510385404354672"
            ],
            [
              1699653778.33,
              "0.12991695859323082"
            ],
            [
              1699653838.33,
              "0.12877932214349952"
            ],
            [
              1699653898.33,
              "0.12309136287256502"
            ],
            [
              1699653958.33,
              "0.08551599049277"
            ],
            [
              1699654018.33,
              "0.13404119475123766"
            ],
            [
              1699654078.33,
              "0.12398079503979959"
            ],
            [
              1699654138.33,
              "0.10434652768345677"
            ],
            [
              1699654198.33,
              "0.1236727546657706"
            ],
            [
              1699654258.33,
              "0.14235003229947296"
            ],
            [
              1699654318.33,
              "0.1314077592597285"
            ],
            [
              1699654378.33,
              "0.11780121069075009"
            ],
            [
              1699654438.33,
              "0.1450755457324068"
            ],
            [
              1699654498.33,
              "0.12377438095155524"
            ],
            [
              1699654558.33,
              "0.12611542228178654"
            ],
            [
              1699654618.33,
              "0.16856694686741952"
            ],
            [
              1699654678.33,
              "0.1337268775863178"
            ],
            [
              1699654738.33,
              "0.1281858352813519"
            ],
            [
              1699654798.33,
              "0.13025363331148404"
            ],
            [
              1699654858.33,
              "0.12895737525974862"
            ],
            [
              1699654918.33,
              "0.10535457036796114"
            ],
            [
              1699654978.33,
              "0.15137183837070894"
            ]
          ],
          "2": [
            [
              1699653178.33,
              "0.11655345028880827"
            ],
            [
              1699653238.33,
              "0.10491374362496625"
            ],
            [
              1699653298.33,
              "0.10514439615479274"
            ],
            [
              1699653358.33,
              "0.1085803691547935"
            ],
            [
              1699653418.33,
              "0.1179256550279297"
            ],
            [
              1699653478.33,
              "0.13261415570030033"
            ],
            [
              1699653538.33,
              "0.11475073810859206"
            ],
            [
              1699653598.33,
              "0.09074644965609963"
            ],
            [
              1699653658.33,
              "0.08667051262201265"
            ],
            [
              1699653718.33,
              "0.08780736166114389"
            ],
            [
              1699653778.33,
              "0.09484241253339991"
            ],
            [
              1699653838.33,
              "0.10249104358757968"
            ],
            [
              1699653898.33,
              "0.10791926675079869"
            ],
            [
              1699653958.33,
              "0.11595323711953302"
            ],
            [
              1699654018.33,
              "0.11805526796859583"
            ],
            [
              1699654078.33,
              "0.1056573548939761"
            ],
            [
              1699654138.33,
              "0.09321170905963694"
            ],
            [
              1699654198.33,
              "0.08964075087079273"
            ],
            [
              1699654258.33,
              "0.08718744824022248"
            ],
            [
              1699654318.33,
              "0.09282011874556267"
            ],
            [
              1699654378.33,
              "0.11866831060925244"
            ],
            [
              1699654438.33,
              "0.10092880410148883"
            ],
            [
              1699654498.33,
              "0.12433722378115156"
            ],
            [
              1699654558.33,
              "0.09245436237760468"
            ],
            [
              1699654618.33,
              "0.10788270542507193"
            ],
            [
              1699654678.33,
              "0.12454508026838387"
            ],
            [
              1699654738.33,
              "0.14882530404586752"
            ],
            [
              1699654798.33,
              "0.0854188463637291"
            ],
            [
              1699654858.33,
              "0.12443356351346466"
            ],
            [
              1699654918.33,
              "0.08057281039219684"
            ],
            [
              1699654978.33,
              "0.09158088688432423"
            ]
          ]
        },
        "incoming_byte_rate": {
          "0": [
            [
              1699653178.33,
              "219.7539159282745"
            ],
            [
              1699653238.33,
              "214.7123612640854"
            ],
            [
              1699653298.33,
              "217.2490738534197"
            ],
            [
              1699653358.33,
              "218.6161314916413"
            ],
            [
              1699653418.33,
              "216.42252029650547"
            ],
            [
              1699653478.33,
              "224.8144145515685"
            ],
            [
              1699653538.33,
              "216.12365373349556"
            ],
            [
              1699653598.33,
              "218"
            ],
            [
              1699653658.33,
              "217.46009906439187"
            ],
            [
              1699653718.33,
              "219.54217593988452"
            ],
            [
              1699653778.33,
              "220.61423785426447"
            ],
            [
              1699653838.33,
              "215.39981650081165"
            ],
            [
              1699653898.33,
              "217.17267552182165"
            ],
            [
              1699653958.33,
              "221.726770320107"
            ],
            [
              1699654018.33,
              "216.76286559635253"
            ],
            [
              1699654078.33,
              "219.28426009958548"
            ],
            [
              1699654138.33,
              "216.38791006997866"
            ],
            [
              1699654198.33,
              "219.44024205748866"
            ],
            [
              1699654258.33,
              "218.92377455469048"
            ],
            [
              1699654318.33,
              "216.53766769865842"
            ],
            [
              1699654378.33,
              "222.752417845079"
            ],
            [
              1699654438.33,
              "217.580481554642"
            ],
            [
              1699654498.33,
              "219.4925493354813"
            ],
            [
              1699654558.33,
              "218.8695628578406"
            ],
            [
              1699654618.33,
              "219.02009358656758"
            ],
            [
              1699654678.33,
              "226.24539892770926"
            ],
            [
              1699654738.33,
              "218.1167385361179"
            ],
            [
              1699654798.33,
              "222.83799985247975"
            ],
            [
              1699654858.33,
              "221.01952694482534"
            ],
            [
              1699654918.33,
              "218.16683346088186"
            ],
            [
              1699654978.33,
              "223.78037910749302"
            ]
          ],
          "1": [
            [
              1699653178.33,
              "383.62867191083774"
            ],
            [
              1699653238.33,
              "384.79947045944436"
            ],
            [
              1699653298.33,
              "385.6965386940722"
            ],
            [
              1699653358.33,
              "383.16646869166806"
            ],
            [
              1699653418.33,
              "385.1742458963521"
            ],
            [
              1699653478.33,
              "385.01511394898614"
            ],
            [
              1699653538.33,
              "383.0104827492621"
            ],
            [
              1699653598.33,
              "385.9710511159816"
            ],
            [
              1699653658.33,
              "383.63518977878965"
            ],
            [
              1699653718.33,
              "391.0874355154672"
            ],
            [
              1699653778.33,
              "384.1573377794413"
            ],
            [
              1699653838.33,
              "382.80269894551253"
            ],
            [
              1699653898.33,
              "385.92105039372757"
            ],
            [
              1699653958.33,
              "385.83414169261926"
            ],
            [
              1699654018.33,
              "386.95928863764857"
            ],
            [
              1699654078.33,
              "384.1638736240366"
            ],
            [
              1699654138.33,
              "383.5635054103178"
            ],
            [
              1699654198.33,
              "389.2687174542995"
            ],
            [
              1699654258.33,
              "384.8382588540069"
            ],
            [
              1699654318.33,
              "386.75643429350606"
            ],
            [
              1699654378.33,
              "385.0088664575092"
            ],
            [
              1699654438.33,
              "388.7837306358207"
            ],
            [
              1699654498.33,
              "388.38394322273433"
            ],
            [
              1699654558.33,
              "385.0876594583532"
            ],
            [
              1699654618.33,
              "387.9266977408668"
            ],
            [
              1699654678.33,
              "387.0262888357801"
            ],
            [
              1699654738.33,
              "385.21905281834313"
            ],
            [
              1699654798.33,
              "388.71632626049916"
            ],
            [
              1699654858.33,
              "385.24534225073364"
            ],
            [
              1699654918.33,
              "387.81326718226694"
            ],
            [
              1699654978.33,
              "383.46797170918114"
            ]
          ],
          "2": [
            [
              1699653178.33,
              "219.71592113631544"
            ],
            [
              1699653238.33,
              "217.34562607204117"
            ],
            [
              1699653298.33,
              "217.183202304377"
            ],
            [
              1699653358.33,
              "218.77885650588675"
            ],
            [
              1699653418.33,
              "216.74768007070261"
            ],
            [
              1699653478.33,
              "218.89179463108252"
            ],
            [
              1699653538.33,
              "219.53482058011676"
            ],
            [
              1699653598.33,
              "216.88317628762238"
            ],
            [
              1699653658.33,
              "221.2333866487169"
            ],
            [
              1699653718.33,
              "220.65268065268066"
            ],
            [
              1699653778.33,
              "221.33371075413393"
            ],
            [
              1699653838.33,
              "220.83350931264795"
            ],
            [
              1699653898.33,
              "217.71146673083865"
            ],
            [
              1699653958.33,
              "223.19074128332844"
            ],
            [
              1699654018.33,
              "218.42066201573402"
            ],
            [
              1699654078.33,
              "220.57502246181494"
            ],
            [
              1699654138.33,
              "219.57213128106085"
            ],
            [
              1699654198.33,
              "221.27021254957793"
            ],
            [
              1699654258.33,
              "221.6410159453061"
            ],
            [
              1699654318.33,
              "217.67772388930672"
            ],
            [
              1699654378.33,
              "216.43935831112503"
            ],
            [
              1699654438.33,
              "221.78091985066325"
            ],
            [
              1699654498.33,
              "218.74080218339444"
            ],
            [
              1699654558.33,
              "219.79174472751495"
            ],
            [
              1699654618.33,
              "218.2391662095447"
            ],
            [
              1699654678.33,
              "220.1275298031605"
            ],
            [
              1699654738.33,
              "216.71193749824985"
            ],
            [
              1699654798.33,
              "220.656889948791"
            ],
            [
              1699654858.33,
              "219.0759137877311"
            ],
            [
              1699654918.33,
              "220.15549361889393"
            ],
            [
              1699654978.33,
              "216.57769869513643"
            ]
          ]
        },
        "memory_usage_bytes": {
          "0": [
            [
              1699653178.33,
              "1885257728"
            ],
            [
              1699653238.33,
              "1885298688"
            ],
            [
              1699653298.33,
              "1885470720"
            ],
            [
              1699653358.33,
              "1885650944"
            ],
            [
              1699653418.33,
              "1886035968"
            ],
            [
              1699653478.33,
              "1886556160"
            ],
            [
              1699653538.33,
              "1886416896"
            ],
            [
              1699653598.33,
              "1886912512"
            ],
            [
              1699653658.33,
              "1887010816"
            ],
            [
              1699653718.33,
              "1886855168"
            ],
            [
              1699653778.33,
              "1887076352"
            ],
            [
              1699653838.33,
              "1887281152"
            ],
            [
              1699653898.33,
              "1887502336"
            ],
            [
              1699653958.33,
              "1888051200"
            ],
            [
              1699654018.33,
              "1888129024"
            ],
            [
              1699654078.33,
              "1888677888"
            ],
            [
              1699654138.33,
              "1885081600"
            ],
            [
              1699654198.33,
              "1885315072"
            ],
            [
              1699654258.33,
              "1885896704"
            ],
            [
              1699654318.33,
              "1886253056"
            ],
            [
              1699654378.33,
              "1886867456"
            ],
            [
              1699654438.33,
              "1886785536"
            ],
            [
              1699654498.33,
              "1887719424"
            ],
            [
              1699654558.33,
              "1887412224"
            ],
            [
              1699654618.33,
              "1887621120"
            ],
            [
              1699654678.33,
              "1887936512"
            ],
            [
              1699654738.33,
              "1887821824"
            ],
            [
              1699654798.33,
              "1887928320"
            ],
            [
              1699654858.33,
              "1888075776"
            ],
            [
              1699654918.33,
              "1888530432"
            ],
            [
              1699654978.33,
              "1888546816"
            ]
          ],
          "1": [
            [
              1699653178.33,
              "1913700352"
            ],
            [
              1699653238.33,
              "1913901056"
            ],
            [
              1699653298.33,
              "1914200064"
            ],
            [
              1699653358.33,
              "1915219968"
            ],
            [
              1699653418.33,
              "1915105280"
            ],
            [
              1699653478.33,
              "1915265024"
            ],
            [
              1699653538.33,
              "1915518976"
            ],
            [
              1699653598.33,
              "1923919872"
            ],
            [
              1699653658.33,
              "1924022272"
            ],
            [
              1699653718.33,
              "1924190208"
            ],
            [
              1699653778.33,
              "1925242880"
            ],
            [
              1699653838.33,
              "1925308416"
            ],
            [
              1699653898.33,
              "1931919360"
            ],
            [
              1699653958.33,
              "1925656576"
            ],
            [
              1699654018.33,
              "1925754880"
            ],
            [
              1699654078.33,
              "1925853184"
            ],
            [
              1699654138.33,
              "1926062080"
            ],
            [
              1699654198.33,
              "1926811648"
            ],
            [
              1699654258.33,
              "1926463488"
            ],
            [
              1699654318.33,
              "1926656000"
            ],
            [
              1699654378.33,
              "1926868992"
            ],
            [
              1699654438.33,
              "1929719808"
            ],
            [
              1699654498.33,
              "1929908224"
            ],
            [
              1699654558.33,
              "1930031104"
            ],
            [
              1699654618.33,
              "1936867328"
            ],
            [
              1699654678.33,
              "1937125376"
            ],
            [
              1699654738.33,
              "1937600512"
            ],
            [
              1699654798.33,
              "1937715200"
            ],
            [
              1699654858.33,
              "1937977344"
            ],
            [
              1699654918.33,
              "1938030592"
            ],
            [
              1699654978.33,
              "1938178048"
            ]
          ],
          "2": [
            [
              1699653178.33,
              "1787752448"
            ],
            [
              1699653238.33,
              "1788223488"
            ],
            [
              1699653298.33,
              "1788121088"
            ],
            [
              1699653358.33,
              "1788305408"
            ],
            [
              1699653418.33,
              "1788489728"
            ],
            [
              1699653478.33,
              "1789276160"
            ],
            [
              1699653538.33,
              "1789046784"
            ],
            [
              1699653598.33,
              "1797337088"
            ],
            [
              1699653658.33,
              "1797836800"
            ],
            [
              1699653718.33,
              "1797746688"
            ],
            [
              1699653778.33,
              "1798590464"
            ],
            [
              1699653838.33,
              "1798328320"
            ],
            [
              1699653898.33,
              "1798729728"
            ],
            [
              1699653958.33,
              "1799311360"
            ],
            [
              1699654018.33,
              "1799708672"
            ],
            [
              1699654078.33,
              "1800224768"
            ],
            [
              1699654138.33,
              "1800654848"
            ],
            [
              1699654198.33,
              "1800527872"
            ],
            [
              1699654258.33,
              "1801375744"
            ],
            [
              1699654318.33,
              "1800937472"
            ],
            [
              1699654378.33,
              "1801052160"
            ],
            [
              1699654438.33,
              "1801113600"
            ],
            [
              1699654498.33,
              "1801220096"
            ],
            [
              1699654558.33,
              "1805484032"
            ],
            [
              1699654618.33,
              "1815597056"
            ],
            [
              1699654678.33,
              "1817792512"
            ],
            [
              1699654738.33,
              "1818116096"
            ],
            [
              1699654798.33,
              "1817919488"
            ],
            [
              1699654858.33,
              "1818333184"
            ],
            [
              1699654918.33,
              "1818365952"
            ],
            [
              1699654978.33,
              "1818927104"
            ]
          ]
        },
        "outgoing_byte_rate": {
          "0": [
            [
              1699653178.33,
              "61.350466112370874"
            ],
            [
              1699653238.33,
              "41.36484940907358"
            ],
            [
              1699653298.33,
              "41.85351270553064"
            ],
            [
              1699653358.33,
              "57.21096605626166"
            ],
            [
              1699653418.33,
              "41.69332245042024"
            ],
            [
              1699653478.33,
              "71.01559996483178"
            ],
            [
              1699653538.33,
              "41.6366258111031"
            ],
            [
              1699653598.33,
              "41.997136558870984"
            ],
            [
              1699653658.33,
              "41.89406099518459"
            ],
            [
              1699653718.33,
              "41.88408822096987"
            ],
            [
              1699653778.33,
              "57.604332972228036"
            ],
            [
              1699653838.33,
              "41.49709473288009"
            ],
            [
              1699653898.33,
              "41.838622456240216"
            ],
            [
              1699653958.33,
              "60.219634630687054"
            ],
            [
              1699654018.33,
              "41.759627509942774"
            ],
            [
              1699654078.33,
              "64.45758014274853"
            ],
            [
              1699654138.33,
              "41.68734491315137"
            ],
            [
              1699654198.33,
              "41.82845040593011"
            ],
            [
              1699654258.33,
              "60.41174258106762"
            ],
            [
              1699654318.33,
              "41.70643006370741"
            ],
            [
              1699654378.33,
              "73.36787516624013"
            ],
            [
              1699654438.33,
              "66.22357463164639"
            ],
            [
              1699654498.33,
              "42.28528472091712"
            ],
            [
              1699654558.33,
              "42.16524216524217"
            ],
            [
              1699654618.33,
              "42.19420896179676"
            ],
            [
              1699654678.33,
              "74.46436416267699"
            ],
            [
              1699654738.33,
              "42.021306577983204"
            ],
            [
              1699654798.33,
              "87.34672790469199"
            ],
            [
              1699654858.33,
              "60.85006517169485"
            ],
            [
              1699654918.33,
              "42.02966800094184"
            ],
            [
              1699654978.33,
              "87.37373737373738"
            ]
          ],
          "1": [
            [
              1699653178.33,
              "53.56535438709562"
            ],
            [
              1699653238.33,
              "53.50883468059232"
            ],
            [
              1699653298.33,
              "75.37732699629436"
            ],
            [
              1699653358.33,
              "53.49926688838823"
            ],
            [
              1699653418.33,
              "70.09730271159762"
            ],
            [
              1699653478.33,
              "53.5442652874962"
            ],
            [
              1699653538.33,
              "53.47727394591503"
            ],
            [
              1699653598.33,
              "75.64106139625109"
            ],
            [
              1699653658.33,
              "53.568231460593694"
            ],
            [
              1699653718.33,
              "84.80651614784725"
            ],
            [
              1699653778.33,
              "53.6402588787895"
            ],
            [
              1699653838.33,
              "53.44575041977779"
            ],
            [
              1699653898.33,
              "70.24178660405427"
            ],
            [
              1699653958.33,
              "53.66046127909896"
            ],
            [
              1699654018.33,
              "76.07871909271839"
            ],
            [
              1699654078.33,
              "53.645067594219526"
            ],
            [
              1699654138.33,
              "53.55960116714104"
            ],
            [
              1699654198.33,
              "76.24488278562991"
            ],
            [
              1699654258.33,
              "53.75011227881075"
            ],
            [
              1699654318.33,
              "70.40330548818827"
            ],
            [
              1699654378.33,
              "53.767498697144504"
            ],
            [
              1699654438.33,
              "80.17706181554854"
            ],
            [
              1699654498.33,
              "85.9988961570394"
            ],
            [
              1699654558.33,
              "53.785863234342415"
            ],
            [
              1699654618.33,
              "76.52415380310305"
            ],
            [
              1699654678.33,
              "53.83521960523238"
            ],
            [
              1699654738.33,
              "53.79940302801453"
            ],
            [
              1699654798.33,
              "86.07251612400454"
            ],
            [
              1699654858.33,
              "53.80327279266319"
            ],
            [
              1699654918.33,
              "69.36119751721051"
            ],
            [
              1699654978.33,
              "53.65053032843302"
            ]
          ],
          "2": [
            [
              1699653178.33,
              "54.699586733071946"
            ],
            [
              1699653238.33,
              "41.87213241284679"
            ],
            [
              1699653298.33,
              "41.84172911162126"
            ],
            [
              1699653358.33,
              "57.25464541070026"
            ],
            [
              1699653418.33,
              "41.7578047325512"
            ],
            [
              1699653478.33,
              "41.77376455600009"
            ],
            [
              1699653538.33,
              "62.78422588746352"
            ],
            [
              1699653598.33,
              "41.782920916617115"
            ],
            [
              1699653658.33,
              "63.88011897708911"
            ],
            [
              1699653718.33,
              "42.09593958323621"
            ],
            [
              1699653778.33,
              "56.423078737557205"
            ],
            [
              1699653838.33,
              "63.66957005381303"
            ],
            [
              1699653898.33,
              "41.94239237673557"
            ],
            [
              1699653958.33,
              "64.67600957077983"
            ],
            [
              1699654018.33,
              "42.07896299228182"
            ],
            [
              1699654078.33,
              "57.06230195364156"
            ],
            [
              1699654138.33,
              "63.82052119263018"
            ],
            [
              1699654198.33,
              "42.17832354511479"
            ],
            [
              1699654258.33,
              "64.06454819440181"
            ],
            [
              1699654318.33,
              "41.93682244929716"
            ],
            [
              1699654378.33,
              "41.69714703730797"
            ],
            [
              1699654438.33,
              "42.257996187248466"
            ],
            [
              1699654498.33,
              "42.14159576175951"
            ],
            [
              1699654558.33,
              "57.442926617896546"
            ],
            [
              1699654618.33,
              "42.04492471407806"
            ],
            [
              1699654678.33,
              "41.91616766467065"
            ],
            [
              1699654738.33,
              "41.75067204301075"
            ],
            [
              1699654798.33,
              "67.900513125873"
            ],
            [
              1699654858.33,
              "57.30267472163564"
            ],
            [
              1699654918.33,
              "41.89280370815853"
            ],
            [
              1699654978.33,
              "41.72474125915602"
            ]
          ]
        },
        "volume_stats_capacity_bytes": {
          "0": [
            [
              1699653178.33,
              "10464022528"
            ],
            [
              1699653238.33,
              "10464022528"
            ],
            [
              1699653298.33,
              "10464022528"
            ],
            [
              1699653358.33,
              "10464022528"
            ],
            [
              1699653418.33,
              "10464022528"
            ],
            [
              1699653478.33,
              "10464022528"
            ],
            [
              1699653538.33,
              "10464022528"
            ],
            [
              1699653598.33,
              "10464022528"
            ],
            [
              1699653658.33,
              "10464022528"
            ],
            [
              1699653718.33,
              "10464022528"
            ],
            [
              1699653778.33,
              "10464022528"
            ],
            [
              1699653838.33,
              "10464022528"
            ],
            [
              1699653898.33,
              "10464022528"
            ],
            [
              1699653958.33,
              "10464022528"
            ],
            [
              1699654018.33,
              "10464022528"
            ],
            [
              1699654078.33,
              "10464022528"
            ],
            [
              1699654138.33,
              "10464022528"
            ],
            [
              1699654198.33,
              "10464022528"
            ],
            [
              1699654258.33,
              "10464022528"
            ],
            [
              1699654318.33,
              "10464022528"
            ],
            [
              1699654378.33,
              "10464022528"
            ],
            [
              1699654438.33,
              "10464022528"
            ],
            [
              1699654498.33,
              "10464022528"
            ],
            [
              1699654558.33,
              "10464022528"
            ],
            [
              1699654618.33,
              "10464022528"
            ],
            [
              1699654678.33,
              "10464022528"
            ],
            [
              1699654738.33,
              "10464022528"
            ],
            [
              1699654798.33,
              "10464022528"
            ],
            [
              1699654858.33,
              "10464022528"
            ],
            [
              1699654918.33,
              "10464022528"
            ],
            [
              1699654978.33,
              "10464022528"
            ]
          ],
          "1": [
            [
              1699653178.33,
              "10464022528"
            ],
            [
              1699653238.33,
              "10464022528"
            ],
            [
              1699653298.33,
              "10464022528"
            ],
            [
              1699653358.33,
              "10464022528"
            ],
            [
              1699653418.33,
              "10464022528"
            ],
            [
              1699653478.33,
              "10464022528"
            ],
            [
              1699653538.33,
              "10464022528"
            ],
            [
              1699653598.33,
              "10464022528"
            ],
            [
              1699653658.33,
              "10464022528"
            ],
            [
              1699653718.33,
              "10464022528"
            ],
            [
              1699653778.33,
              "10464022528"
            ],
            [
              1699653838.33,
              "10464022528"
            ],
            [
              1699653898.33,
              "10464022528"
            ],
            [
              1699653958.33,
              "10464022528"
            ],
            [
              1699654018.33,
              "10464022528"
            ],
            [
              1699654078.33,
              "10464022528"
            ],
            [
              1699654138.33,
              "10464022528"
            ],
            [
              1699654198.33,
              "10464022528"
            ],
            [
              1699654258.33,
              "10464022528"
            ],
            [
              1699654318.33,
              "10464022528"
            ],
            [
              1699654378.33,
              "10464022528"
            ],
            [
              1699654438.33,
              "10464022528"
            ],
            [
              1699654498.33,
              "10464022528"
            ],
            [
              1699654558.33,
              "10464022528"
            ],
            [
              1699654618.33,
              "10464022528"
            ],
            [
              1699654678.33,
              "10464022528"
            ],
            [
              1699654738.33,
              "10464022528"
            ],
            [
              1699654798.33,
              "10464022528"
            ],
            [
              1699654858.33,
              "10464022528"
            ],
            [
              1699654918.33,
              "10464022528"
            ],
            [
              1699654978.33,
              "10464022528"
            ]
          ],
          "2": [
            [
              1699653178.33,
              "10464022528"
            ],
            [
              1699653238.33,
              "10464022528"
            ],
            [
              1699653298.33,
              "10464022528"
            ],
            [
              1699653358.33,
              "10464022528"
            ],
            [
              1699653418.33,
              "10464022528"
            ],
            [
              1699653478.33,
              "10464022528"
            ],
            [
              1699653538.33,
              "10464022528"
            ],
            [
              1699653598.33,
              "10464022528"
            ],
            [
              1699653658.33,
              "10464022528"
            ],
            [
              1699653718.33,
              "10464022528"
            ],
            [
              1699653778.33,
              "10464022528"
            ],
            [
              1699653838.33,
              "10464022528"
            ],
            [
              1699653898.33,
              "10464022528"
            ],
            [
              1699653958.33,
              "10464022528"
            ],
            [
              1699654018.33,
              "10464022528"
            ],
            [
              1699654078.33,
              "10464022528"
            ],
            [
              1699654138.33,
              "10464022528"
            ],
            [
              1699654198.33,
              "10464022528"
            ],
            [
              1699654258.33,
              "10464022528"
            ],
            [
              1699654318.33,
              "10464022528"
            ],
            [
              1699654378.33,
              "10464022528"
            ],
            [
              1699654438.33,
              "10464022528"
            ],
            [
              1699654498.33,
              "10464022528"
            ],
            [
              1699654558.33,
              "10464022528"
            ],
            [
              1699654618.33,
              "10464022528"
            ],
            [
              1699654678.33,
              "10464022528"
            ],
            [
              1699654738.33,
              "10464022528"
            ],
            [
              1699654798.33,
              "10464022528"
            ],
            [
              1699654858.33,
              "10464022528"
            ],
            [
              1699654918.33,
              "10464022528"
            ],
            [
              1699654978.33,
              "10464022528"
            ]
          ]
        },
        "volume_stats_used_bytes": {
          "0": [
            [
              1699653178.33,
              "48295936"
            ],
            [
              1699653238.33,
              "48295936"
            ],
            [
              1699653298.33,
              "48328704"
            ],
            [
              1699653358.33,
              "48349184"
            ],
            [
              1699653418.33,
              "48349184"
            ],
            [
              1699653478.33,
              "48377856"
            ],
            [
              1699653538.33,
              "48398336"
            ],
            [
              1699653598.33,
              "48418816"
            ],
            [
              1699653658.33,
              "48418816"
            ],
            [
              1699653718.33,
              "48451584"
            ],
            [
              1699653778.33,
              "48472064"
            ],
            [
              1699653838.33,
              "48496640"
            ],
            [
              1699653898.33,
              "48496640"
            ],
            [
              1699653958.33,
              "48525312"
            ],
            [
              1699654018.33,
              "48525312"
            ],
            [
              1699654078.33,
              "48549888"
            ],
            [
              1699654138.33,
              "48574464"
            ],
            [
              1699654198.33,
              "48599040"
            ],
            [
              1699654258.33,
              "48599040"
            ],
            [
              1699654318.33,
              "48619520"
            ],
            [
              1699654378.33,
              "48640000"
            ],
            [
              1699654438.33,
              "48656384"
            ],
            [
              1699654498.33,
              "48676864"
            ],
            [
              1699654558.33,
              "48676864"
            ],
            [
              1699654618.33,
              "48709632"
            ],
            [
              1699654678.33,
              "48709632"
            ],
            [
              1699654738.33,
              "48738304"
            ],
            [
              1699654798.33,
              "48738304"
            ],
            [
              1699654858.33,
              "48775168"
            ],
            [
              1699654918.33,
              "48775168"
            ],
            [
              1699654978.33,
              "48807936"
            ]
          ],
          "1": [
            [
              1699653178.33,
              "48369664"
            ],
            [
              1699653238.33,
              "48369664"
            ],
            [
              1699653298.33,
              "48402432"
            ],
            [
              1699653358.33,
              "48402432"
            ],
            [
              1699653418.33,
              "48435200"
            ],
            [
              1699653478.33,
              "48435200"
            ],
            [
              1699653538.33,
              "48455680"
            ],
            [
              1699653598.33,
              "48480256"
            ],
            [
              1699653658.33,
              "48480256"
            ],
            [
              1699653718.33,
              "48513024"
            ],
            [
              1699653778.33,
              "48513024"
            ],
            [
              1699653838.33,
              "48545792"
            ],
            [
              1699653898.33,
              "48566272"
            ],
            [
              1699653958.33,
              "48590848"
            ],
            [
              1699654018.33,
              "48611328"
            ],
            [
              1699654078.33,
              "48611328"
            ],
            [
              1699654138.33,
              "48635904"
            ],
            [
              1699654198.33,
              "48656384"
            ],
            [
              1699654258.33,
              "48676864"
            ],
            [
              1699654318.33,
              "48676864"
            ],
            [
              1699654378.33,
              "48701440"
            ],
            [
              1699654438.33,
              "48734208"
            ],
            [
              1699654498.33,
              "48734208"
            ],
            [
              1699654558.33,
              "48754688"
            ],
            [
              1699654618.33,
              "48754688"
            ],
            [
              1699654678.33,
              "48787456"
            ],
            [
              1699654738.33,
              "48787456"
            ],
            [
              1699654798.33,
              "48820224"
            ],
            [
              1699654858.33,
              "48848896"
            ],
            [
              1699654918.33,
              "48848896"
            ],
            [
              1699654978.33,
              "48873472"
            ]
          ],
          "2": [
            [
              1699653178.33,
              "48320512"
            ],
            [
              1699653238.33,
              "48320512"
            ],
            [
              1699653298.33,
              "48353280"
            ],
            [
              1699653358.33,
              "48353280"
            ],
            [
              1699653418.33,
              "48381952"
            ],
            [
              1699653478.33,
              "48414720"
            ],
            [
              1699653538.33,
              "48414720"
            ],
            [
              1699653598.33,
              "48435200"
            ],
            [
              1699653658.33,
              "48467968"
            ],
            [
              1699653718.33,
              "48467968"
            ],
            [
              1699653778.33,
              "48500736"
            ],
            [
              1699653838.33,
              "48500736"
            ],
            [
              1699653898.33,
              "48525312"
            ],
            [
              1699653958.33,
              "48549888"
            ],
            [
              1699654018.33,
              "48570368"
            ],
            [
              1699654078.33,
              "48570368"
            ],
            [
              1699654138.33,
              "48599040"
            ],
            [
              1699654198.33,
              "48599040"
            ],
            [
              1699654258.33,
              "48627712"
            ],
            [
              1699654318.33,
              "48652288"
            ],
            [
              1699654378.33,
              "48672768"
            ],
            [
              1699654438.33,
              "48672768"
            ],
            [
              1699654498.33,
              "48701440"
            ],
            [
              1699654558.33,
              "48701440"
            ],
            [
              1699654618.33,
              "48734208"
            ],
            [
              1699654678.33,
              "48762880"
            ],
            [
              1699654738.33,
              "48762880"
            ],
            [
              1699654798.33,
              "48795648"
            ],
            [
              1699654858.33,
              "48795648"
            ],
            [
              1699654918.33,
              "48828416"
            ],
            [
              1699654978.33,
              "48828416"
            ]
          ]
        }
      }
    }
  }
}
```
</p></details>

MikeEdgar commented:
@riccardo-forina , does the shape of the metrics in the response (example in the PR description) seem usable to you? The `metrics` attribute is a field that may optionally be fetched for a `describeCluster` operation and is not available for `listClusters`.

riccardo-forina commented:
Yeah it looks ok. Why is the timestamp a float? I suppose I can't test this in advance as I need a Kafka cluster set up to store metrics on a Prometheus instance, right?

MikeEdgar commented:
> Why is the timestamp a float? 

It's the epoch timestamp as received from Prometheus. It's definitely inconsistent from the rest of the API, so I'm glad you mentioned it. I'll change to an ISO timestamp.

> I suppose I can't test this in advance as I need a Kafka cluster set up to store metrics on a Prometheus instance, right?

Prometheus has been deployed to the demo environment, I will ping you the endpoint to configure in your `.env`.

MikeEdgar commented:
@riccardo-forina , I've updated this PR again to add in "value" metrics that are a single value (as opposed to a range over time). The sample data in the description is also updated to reflect the current state. Each of the value and range metrics are an array rather than being keyed by something, like nodeId, as was the case earlier. This should allow the same structure to be reused for other metrics. Each entry in the array could potentially have additional attributes besides the value or range, such as the nodeId for the per-broker metrics.

MikeEdgar commented:
@daveconde , the example resources added in this PR to deploy Prometheus and its configuration assume the Prometheus operator is already present in the k8s cluster. Please take a look at what is added here and let me know if you spot any issues.

@katheris , @tombentley , feedback would be appreciated on the approach to gather metrics (especially the selection of metrics themselves) for the Kafka cluster overview. The Prometheus queries are separated into two template files added in this PR, one for ranges and the other for instant values.

Instant values:
- `kafka_server_kafkaserver_brokerstate` should be used to determine the number of brokers "up". The values can be found in `org.apache.kafka.metadata.BrokerState` and I believe anything other than `3` should mean a broker is offline.
- `kafka_controller_kafkacontroller_globaltopiccount` - count of topics in the cluster
- `kafka_controller_kafkacontroller_globalpartitioncount` - count of partitions in the cluster
- `kafka_cluster_partition_underreplicated` - count of partitions that are in an "under replicated" state

Range values:
- `container_cpu_usage_seconds_total` - for the CPU usage of each broker pod over time
- `container_memory_usage_bytes` - for the memory usage of each broker pod over time
- `kafka_server_socket_server_metrics_incoming_byte_rate` - incoming bytes through the listener published by the API (either auto-selected or via annotation set by an admin)
- `kafka_server_socket_server_metrics_outgoing_byte_rate` - outgoing bytes through the listener published by the API (either auto-selected or via annotation set by an admin)
- `kubelet_volume_stats_capacity_bytes` - per broker PVC capacity
- `kubelet_volume_stats_used_bytes` - per broker PVC usage

MikeEdgar commented on api/src/main/java/com/github/eyefloaters/console/api/service/MetricsService.java:
```diff
@@ -0,0 +1,125 @@
+
+    CompletionStage<Map<String, List<Metrics.RangeMetric>>> queryRanges(String query) {
+        Instant now = Instant.now();
+        String start = Double.toString(now.minus(30, ChronoUnit.MINUTES).toEpochMilli() / 1000d);
```
Metrics window is currently hard-coded to 30 minutes, could be made configurable if necessary.

MikeEdgar commented:
@daveconde , @katheris , @tombentley - bump. Please see my [previous comment](https://github.com/eyefloaters/console/pull/171#issuecomment-1828835205).

katheris commented:
@MikeEdgar those metrics look alright to me for the overview page at least. Where are the names coming from? Is that what they get exposed as by default? I just noticed they were different from what appears in the example prometheus and Grafana files on the Strimzi website.

The only other metrics we might need are more around the partition status, e.g. offline partitions or number of partitions below min ISR

MikeEdgar commented:
Thanks @katheris . The metric configuration that I used comes from Strimzi [1]. Which config are you comparing with?

I'll add the additional partition status metrics in the PR for topic status. 

[1] https://github.com/strimzi/strimzi-kafka-operator/blob/b7a1d5225420f05ec1c2c2a8e5c44434cce6ccae/examples/metrics/kafka-metrics.yaml#L69

MikeEdgar commented:
Here's one that may be jumped out as different: `kafka_cluster_partition_underreplicated` versus `kafka_server_replicamanager_underreplicatedpartitions` in the Grafana dashboard. The `kafka_cluster_partition_underreplicated` is broken out by topic which allows the cluster summary to give a count by topic (which is what we have in the mocks).

That may ultimately be a problem if we are unable to determine a metric to show offline partitions in terms of topic counts, which I now suspect we may not...
